### PR TITLE
Handlers API Implementation

### PR DESCRIPTION
**Pull request by stephen-pl**
Standard health check implementation. The helper functions will be used on later pull requests to return the responses. Also added code to start up the server and initialize the db. Verified health check succeeded in postman.

GET http://localhost:3000/api/v1/status

{
    "request_uuid": "EAcPBpwqYbLNeVCEXINKTDjHwbqXLQxe",
    "response_timestamp": "2023-12-05T23:43:15.96430393-08:00",
    "response_state_epoch": 1701848595964,
    "response_code": 200,
    "response": "Healthy"
}

### CODE REVIEW

File: api/v1/api.go [hunk 1]
```diff
@@ -62,3 +66,41 @@ func (s *apiV1) RegisterRoutes(e *echo.Echo) {
...
 }
+func GetTenantId(c echo.Context) int {
+	return int(c.Get(TENANT_CONTEXT).(AuthContext).TenantID)
+}
+
+func CreateErrorResponseEnvelop(c echo.Context, err string) ResponseEnvelope {
+	return ResponseEnvelope{
+		RequestUUID:        c.Response().Header().Get(echo.HeaderXRequestID),
+		ResponseTime:       time.Now(),
+		ResponseStateEpoch: time.Now().UTC().UnixMilli(),
```
Comment (stephen-pl) [inline api/v1/api.go:77]: @jcace what did you intend for this to be?

File: api/v1/api.go [hunk 2]
```diff
@@ -28,9 +33,8 @@ func NewApiV1() *apiV1 {
 // @in header
 // @name Authorization
 func (s *apiV1) RegisterRoutes(e *echo.Echo) {
-	e.GET("/status", s.handleStatus)
-
 	e.Use(AuthMiddleware)
+	e.GET("/status", s.handleStatus)
```
Comment (jcace) [inline api/v1/api.go:37]: I was intending to make this route un-authenticated by putting it before `e.Use(AuthMiddleware)`, but I didn't test if it worked that way - any idea?

File: api/v1/api.go (hunk 1 above)
Comment (jcace) [reply api/v1/api.go:77]: I'm not sure, @ribasushi  what is this?

File: api/v1/api.go (hunk 2 above)
Comment (stephen-pl) [reply api/v1/api.go:37]: I'm pretty sure that would work. We do open ourselves up to anyone who wants to contentiously hit the endpoint since we don't have throttling policies in place. I doubt it is an issue at this point.

File: api/v1/api.go (hunk 1 above)
Comment (jcace) [reply api/v1/api.go:77]: It should be the current Filecoin epoch (which is just `uint` not a `Time` value.  It should look like a number on this thing: https://www.epochclock.io/

You can calculate it like 
```go
func UnixToFilEpoch(unixEpoch int64) int64 {
	return int64(math.Floor(float64(unixEpoch-FILECOIN_GENESIS_UNIX_EPOCH) / 30))
}
```

Where 
`FILECOIN_GENESIS_UNIX_EPOCH = 1598306400`

@ribasushi please chime in if there's any issues with that approach (i.e, should we query an actual chain endpoint for it vs. computing here?) This is how we did it in Estuary

File: api/v1/api.go [hunk 3]
```diff
@@ -62,3 +66,41 @@ func (s *apiV1) RegisterRoutes(e *echo.Echo) {
...
+
+func CreateErrorResponseEnvelop(c echo.Context, err string) ResponseEnvelope {
+	return ResponseEnvelope{
+		RequestUUID:        c.Response().Header().Get(echo.HeaderXRequestID),
+		ResponseTime:       time.Now(),
+		ResponseStateEpoch: time.Now().UTC().UnixMilli(),
+		ResponseCode:       http.StatusInternalServerError,
+		ErrCode:            http.StatusInternalServerError,
+		ErrSlug:            err,
+		Response:           err,
```
Comment (jcace) [inline api/v1/api.go:81]: this should be set to `nil` for the error case

File: initializers/loadEnvVariables.go [hunk 4]
```diff
@@ -0,0 +1,14 @@
+package initializers
```
[the remark this answers is not in the document]
Comment (jcace) [reply initializers/loadEnvVariables.go:1]: Plus you can also do some validation in that InitConfig() too.

File: api/v1/api.go [hunk 5]
```diff
@@ -62,3 +66,41 @@ func (s *apiV1) RegisterRoutes(e *echo.Echo) {
...
+}
+
+func CreateErrorResponseEnvelop(c echo.Context, err string) ResponseEnvelope {
+	return ResponseEnvelope{
+		RequestUUID:        c.Response().Header().Get(echo.HeaderXRequestID),
+		ResponseTime:       time.Now(),
+		ResponseStateEpoch: time.Now().UTC().UnixMilli(),
+		ResponseCode:       http.StatusInternalServerError,
+		ErrCode:            http.StatusInternalServerError,
+		ErrSlug:            err,
```
[the remark this answers is not in the document]
Comment (stephen-pl) [reply api/v1/api.go:80]: Are you saying that errorSlug should error.body? Not sure Im following he error code to error slug map comment.

File: db/main.go [hunk 6]
```diff
@@ -9,8 +13,30 @@ import (
...
 var (
 	log = logging.Logger("router")
+	DB  *gorm.DB
 )
 
+func init() {
+	initializers.LoadEnvVariables()
+}
+
+func ConnectToDB() {
```
[the remark this answers is not in the document]
Comment (stephen-pl) [reply db/main.go:23]: I have now. I would like the next few prs to focus on implementation and ill come back and clean things up

File: config/config.go [hunk 7]
```diff
@@ -0,0 +1,30 @@
+package config
+
+import (
+	"github.com/caarlos0/env/v6"
+	"github.com/joho/godotenv"
+	"github.com/labstack/gommon/log"
+)
+
+type DeltaConfig struct {
```
Comment (jcace) [inline config/config.go:9]: rename `TenantServiceConfig`

File: config/config.go [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+	Settings struct {
```
Comment (jcace) [inline config/config.go:10]: Probably don't need this `Settings` sub struct in here

File: db/main.go [hunk 9]
```diff
@@ -9,15 +10,36 @@ import (
...
+	DB  *gorm.DB
 )
 
+type DeltaDM struct {
+	DB         *gorm.DB
+	DryRunMode bool
+	Config     config.DeltaConfig
+}
+
+func NewDeltaDMDm() *DeltaDM {
```
Comment (jcace) [inline db/main.go:22]: Replace the names, i.e `DeltaDM` -> `SpdTenantSvc` 
`NewDeltaDm` -> `NewSpdTenantSvc`

File: db/main.go [hunk 10]
```diff
@@ -9,15 +10,36 @@ import (
...
+	config := config.InitConfig()
+	if config.Settings.DEBUG {
+		logging.SetDebugLogging()
+	}
+
+	dbi, err := OpenDatabase(config.Settings.DB_URL, config.Settings.DEBUG, config.Settings.DRY_RUN)
+	if err != nil {
+		log.Fatalf("could not connect to delta api: %s", err)
```
Comment (jcace) [inline db/main.go:30]: fix this log message `could not open db`

File: api/v1/api.go [hunk 11]
```diff
@@ -27,38 +36,105 @@ func NewApiV1() *apiV1 {
...
+	case 1:
+		return "error_1"
+	case 2:
+		return "error_2"
+	default:
+		return "unknown"
+	}
+}
+
+func CreateErrorResponseEnvelop(c echo.Context, errorCode int, err string) ResponseEnvelope {
```
Comment (jcace) [inline api/v1/api.go:107]: spelling `Envelop` -> `Envelope`

File: api/v1/api.go [hunk 12]
```diff
@@ -27,38 +36,105 @@ func NewApiV1() *apiV1 {
 // @type apiKey
 // @in header
 // @name Authorization
-func (s *apiV1) RegisterRoutes(e *echo.Echo) {
+func (s *apiV1) RegisterRoutes(e *echo.Echo, config config.DeltaConfig) {
```
Comment (jcace) [inline api/v1/api.go:39]: A few tweaks here - basically, can we clean up so it looks similar to the way we set up the routes in `Delta-dm`? 
i.e here
https://github.com/application-research/delta-dm/blob/main/api/router.go#L47

- remove the `apiV1` struct altogether, don't think it's needed 
- create an `echo` `apiGroup := e.Group("/api/v1")` and put all the routes under that 
- Move all the other middleware in here (ex `e.Use(middleware.RequestID())`)

File: api/v1/api.go (hunk 11 above)
Comment (stephen-pl) [reply api/v1/api.go:107]: good catch

File: api/v1/api.go (hunk 12 above)
Comment (stephen-pl) [reply api/v1/api.go:39]: In the follow up prs I was going to move these functions into their right classes. This was more of skeleton set up for later so things could compile and the general layout would be there. I can move the middleware and api droups in this pr though

**Comment by stephen-pl** [comment]
Current dependencies on/for this PR:
* main
  * **PR #8** <a href="https://app.graphite.dev/github/pr/data-preservation-programs/spade-tenant/8?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/data-preservation-programs/spade-tenant/8?utm_source=stack-comment).

File: api/v1/api.go (hunk 1 above)
Comment (ribasushi) [reply api/v1/api.go]: This is correct, only thing I would change is not using floting point math: integer division is enough. Some SQL examples:

```
-- epoch_from_ts
CREATE OR REPLACE
  FUNCTION epoch_from_ts(TIMESTAMP WITH TIME ZONE) RETURNS INTEGER
LANGUAGE sql PARALLEL SAFE IMMUTABLE STRICT AS $$

  SELECT ( EXTRACT( EPOCH FROM $1 )::BIGINT - 1598306400 ) / 30

$$;

-- ts_from_epoch
CREATE OR REPLACE
  FUNCTION ts_from_epoch(INTEGER) RETURNS TIMESTAMP WITH TIME ZONE
LANGUAGE sql PARALLEL SAFE IMMUTABLE STRICT AS $$

  SELECT TIMEZONE( 'UTC', TO_TIMESTAMP( $1 * 30::BIGINT + 1598306400 ) )

$$;

-- round_down_epoch
-- 2h +/- because network started at 22:00 UTC
CREATE OR REPLACE
  FUNCTION round_down_epoch(INTEGER) RETURNS INTEGER
LANGUAGE sql PARALLEL SAFE IMMUTABLE STRICT AS $$

  SELECT ($1 - 240) / 2880 * 2880 + 240

$$;
```

File: api/v1/api.go [hunk 13]
```diff
@@ -27,38 +29,64 @@ func NewApiV1() *apiV1 {
...
+}
+
+func GetTenantId(c echo.Context) int {
+	return int(c.Get(TENANT_CONTEXT).(AuthContext).TenantID)
+}
 
-	// /storage-contract
-	e.GET("/storage-contract", s.handleGetStorageContract)
-	e.POST("/storage-contract", s.handleSetStorageContract)
+func UnixToFilEpoch(unixEpoch int64) int64 {
```
Comment (jcace) [inline api/v1/api.go:44]: its fine for now since its just one function, but in future it would be good to break this out into a separate sub-package in the repo called `utils` or similar

File: api/v1/handlers.go [hunk 14]
```diff
@@ -2,24 +2,11 @@ package api
```
Comment (jcace) [inline api/v1/handlers.go:2]: rename this file from `handlers.go` to `status.go` now that its only doing one thing

File: main.go [hunk 15]
```diff
@@ -1,10 +1,14 @@
...
+	"github.com/data-preservation-programs/spade-tenant/api/v1"
+	"github.com/data-preservation-programs/spade-tenant/config"
+	"github.com/labstack/echo/v4"
+)
 
 func main() {
-	dbDsn := "postgres://postgres:password@localhost:5432/spade-tenant"
-	debug := true
-
-	db.OpenDatabase(dbDsn, debug)
```
Comment (jcace) [inline main.go:9]: since you removed this, where does the DB get setup now? or will this be in a separate PR?

**Review by jcace — APPROVED** [review]
Looks good, just a few small tweaks left!

File: db/main.go [hunk 16]
```diff
@@ -9,15 +10,36 @@ import (
...
+	DB  *gorm.DB
 )
 
+type SpdTenantSvc struct {
+	DB         *gorm.DB
+	DryRunMode bool
+	Config     config.TenantServiceConfig
+}
+
+func NewSpdTenantSvc() *SpdTenantSvc {
```
Comment (jcace) [inline db/main.go:22]: It's strange to me that this function is inside the `db` subpackage, IMO it should live elsewhere, as it contains more than just the DB struct but also the config (and may have more things added to it in the future). It's more like a global app context object. 

In the past I've put that in a subpackage called `core`, i.e, at the top level directory `/core/tenant-svc.go`

Comment (jcace) [reply db/main.go:22]: In this `db` package, you should just have the stuff that opens the database and nothing to do with InitConfig parsing etc
